### PR TITLE
fix(override): relative file in workspace

### PR DESCRIPTION
**Pull request by KSXGitHub**

### CODE REVIEW

File: hooks/read-package-hook/src/createVersionsOverrider.ts [hunk 1]
```diff
@@ -103,10 +111,14 @@ function overrideDeps (
       )}`
       continue
     }
-    if (versionOverride.linkFileTarget) {
-      deps[
-        versionOverride.targetPkg.name
-      ] = `file:${versionOverride.linkFileTarget}`
+    if (versionOverride.linkFileTarget && dir) {
```
Comment (zkochan) [inline hooks/read-package-hook/src/createVersionsOverrider.ts]: What about when dir is undefined? I assume that would happen when the override is applied on a dependency. Probably an absolute path should be used in that case.

Comment (KSXGitHub) [reply hooks/read-package-hook/src/createVersionsOverrider.ts]: https://github.com/pnpm/pnpm/pull/8055/commits/f1023f69fb335235a920c27f379b88408946b5c9

File: hooks/read-package-hook/src/createVersionsOverrider.ts [hunk 2]
```diff
@@ -20,9 +20,12 @@ export function createVersionsOverrider (
           const pkgPath = override.newPref.substring(5)
           linkTarget = path.isAbsolute(pkgPath) ? pkgPath : path.join(rootDir, pkgPath)
```
Comment (zkochan) [inline hooks/read-package-hook/src/createVersionsOverrider.ts]: what about this one? Should probably follow the same rules. Let's refactor both

Comment (KSXGitHub) [reply hooks/read-package-hook/src/createVersionsOverrider.ts]: Right now, `link:` converts all absolute path to relative path, but `file:` preserves absolute path. The change you requested would alter one of them. Do you still want to change? If you do, do you prefer both converts absolute path to relative or both preserves absolute path?

Comment (zkochan) [reply hooks/read-package-hook/src/createVersionsOverrider.ts]: Isn't this change making file preserve related? The new way `file:` works should be fine for `link:` too.

Comment (KSXGitHub) [reply hooks/read-package-hook/src/createVersionsOverrider.ts]: I have made `link:` behaves like `file:` (preserve absolute path)

**Comment by zkochan** [comment]
We need to verify if the overrides really work after this change. Both on workspace projects and subdeps.

**Comment by KSXGitHub** [comment]
> We need to verify if the overrides really work after this change. Both on workspace projects and subdeps.

Should this verification be in an integration test?

**Comment by zkochan** [comment]
yes, I guess that would be good. It would have to be 8 cases. relative, absolute X project, dependency X file, link

File: pnpm/test/install/overrides.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+        specifier: 'link:../../overrides/pkg',
+        version: 'link:../../overrides/pkg',
+      },
+      'absolute-link-pkg': {
+        specifier: `link:${path.resolve('overrides/pkg')}`,
+        version: 'link:../../overrides/pkg',
+      },
+    },
+  } as typeof lockfile['importers'][string])
+})
```
Comment (KSXGitHub) [inline pnpm/test/install/overrides.ts]: @zkochan So I have written 4 test cases (the "dependency" cases require modifying `registry-mock` and shall be done later). And I see that the `version` fields in the lockfile are probably not correct. What do you think should be the correct values? Should the versions of absolute paths also absolute? And is it intended that `link:` doesn't have `pkg@` like `file:`?

Comment (zkochan) [reply pnpm/test/install/overrides.ts]: > And is it intended that link: doesn't have pkg@ like file:?

Yes, but `file:` also shouldn't have `pkg@` if `pkg` is the real name in the `package.json` of `pkg`.

> Should the versions of absolute paths also absolute?

That should be OK to keep it absolute. absolute overrides doesn't makes sense in any case. I am not sure how users use them. Probably they don't commit those changes.

File: pnpm/test/install/overrides.ts
Comment (zkochan) [inline pnpm/test/install/overrides.ts:1]: move the test to "@pnpm/core"

Comment (KSXGitHub) [reply pnpm/test/install/overrides.ts:1]: https://github.com/pnpm/pnpm/pull/8055/commits/3f349dee26ffac897f310e5412f2ab01c7fb7e1a

**Comment by gluxon** [comment]
Thanks for this fix!

**Review by zkochan — APPROVED** [review]

**Pull request merged**